docs: add design for enforcing host network. (#170)
ceph/ceph-csi-operator: docs/design/hostNetwork.md

# Enabling host networking for controller plugin pods

By default, the Ceph-CSI controller plugins operate on the pod network
but under some circumstances, like setups with a dedicated storage network, 
where the pod network cannot connect to the ceph cluster,
it is necessary to run the Ceph-CSI controller plugin pods on the host network.

This document describes how the ceph-csi-operator can be configured to enforce the use of host network
in the Ceph-CSI controller plugin pods

> The use of host networking can be enabled for a driver's controller plugin by setting `hostNetwork` to `true`in the `ControllerPlugin` section of the corresponding Driver CR.

Reviewer: ```suggestion
The use of host networking can be enabled for a driver's controller plugin by setting `hostNetwork` to `true`in the `ControllerPlugin` section of the corresponding `Driver` CR.
```

> The `useHostNetworking` setting is also available in the `driverSpecDefaults.controllerPlugin` section

Reviewer: `useHostNetworking` to `hostNetwork`

Author: done

of the `OperatorConfig` CR. As this is a default for all Ceph-CSI controller plugins created by the operator, the setting
in concrete Driver CRs will take precedence.

> There is currently no means of enforcing the use of host networking on all controller plugins against driver CR settings.

Reviewer: ```suggestion
There is currently no means of enforcing the use of host networking on all controller plugins against `Driver` CR settings.
```

Reviewer: @obnoxxx this is not done, can you please check

Example:

## OperatorConfig CR

```yaml
kind: OperatorConfig 
apiVersion: csi.ceph.io/v1alpha1
metadata:
  name: ceph-csi-operator-config
  namespace: <operator-namespace>
spec:
   driverSpecDefaults:
     controllerPlugin:
       hostNetwork: true
```
## Driver CR

> ```yaml

Reviewer: a new line is missing here between line 34 and 35

Author: done

> # Enabling  host networking for controller plugin pods

Reviewer: ```suggestion
# Enabling host networking for controller plugin pods
```

By default, the Ceph-CSI controller plugins operate on the pod network
but under some circumstances, like setups with a dedicated storage network, 
where the pod network cannot connect to the ceph cluster,
it is necessary to run the Ceph-CSI controller plugin pods on the host network.

This document describes how the ceph-csi-operator can be configured to enforce the use of host network
in the Ceph-CSI controller plugin pods

> 

Reviewer: Remove extra new lines

Author: done

The use of host networking can be enabled for a driver's controller plugin by setting `hostNetwork` to `true`in the `ControllerPlugin` section of the corresponding Driver CR.

The `useHostNetworking` setting is also available in the `driverSpecDefaults.controllerPlugin` section
of the `OperatorConfig` CR. As this is a default for all Ceph-CSI controller plugins created by the operator, the setting
in concrete Driver CRs will take precedence.

There is currently no means of enforcing the use of host networking on all controller plugins against driver CR settings.

> 

Reviewer: Remove extra new line

Author: done

Example:

> 

Reviewer: Remove extra new line

Author: done

## OperatorConfig CR

kind: OperatorConfig 
apiVersion: csi.ceph.io/v1alpha1
metadata:
  name: ceph-csi-operator-config
  namespace: <operator-namespace>
spec:
   driverSpecDefaults:
     controllerPlugin:

>        hostNetwork: true

Reviewer: place it inside yaml markdown format?

Author: done

## Driver CR
apiVersion: csi.ceph.io/v1alpha1
kind: Driver
metadata:
  name: rbd.csi.ceph.com
  namespace: <operator-namespace> 
spec:
   controllerPlugin:

>      hostNetwork: false

Reviewer: place it inside yaml markdown format?

# Enabling  host networking for controller plugin pods

> By default, the ceph-csi controller plugins operate on the pod network

Reviewer: `ceph-csi` to `Ceph-CSI` lets use the same convention

Author: done

but under some circumstances, like setups with a dedicated storage network, 
where the pod network cannot connect to the ceph cluster,
it is necessary to run the Ceph-CSI controller plugin pods on the host network.

This document describes how the ceph-csi-operator can be configured to enforce the use of host network in 
the Ceph-CSI controller plugin pods

> The use of host networking can be enabled for a driver's controller plugin by setting `useHostNetworking` to `true`in the `ControllerPlugin` section  of the corresponding Driver CR.

Reviewer: ```suggestion
The use of host networking can be enabled for a driver's controller plugin by setting `useHostNetworking` to `true`in the `ControllerPlugin` section of the corresponding Driver CR.
```

> The `useHostNetworking` setting is also available in  the  `driverSpecDefaults.controllerPlugin` section
> of the `OperatorConfig` CR.  As this is a default for all Drivers created by the operator, the setting
> in concrete Driver CRs will take precedence.

Reviewer: ```suggestion
The `useHostNetworking` setting is also available in the `driverSpecDefaults.controllerPlugin` section
of the `OperatorConfig` CR. As this is a default for all Ceph-CSI controller plugins created by the operator, the setting
in concrete Driver CRs will take precedence.
```

> There is currently no means of enforcing the use of host networking on all controller olgins against driver CR settings.

Reviewer 2: ```suggestion
There is currently no means of enforcing the use of host networking on all controller plugins against driver CR settings.
```
😃

Example:

> ```yaml
> kind: OperatorConfig 
> apiVersion: csi.ceph.io/v1alpha1
> 
> 
> metadata:
>   name: ceph-csi-operator-config
>   namespace: <operator-namespace>
> spec:
>   driverSpecDefaults:
>      controllerPlugin:
>          useHostNetworking: true
> ---
> apiVersion: csi.ceph.io/v1alpha1
> kind: Driver
>      controllerPlugin:
>          useHostNetworking: false
> metadata:
>   name: rbd.csi.ceph.com
>   namespace: <operator-namespace> 
> ...
> 
> ```

Reviewer: ```suggestion

## OperatorConfig CR

kind: OperatorConfig 
apiVersion: csi.ceph.io/v1alpha1
metadata:
  name: ceph-csi-operator-config
  namespace: <operator-namespace>
spec:
  driverSpecDefaults:
    controllerPlugin:
      hostNetwork: true

## Driver CR
apiVersion: csi.ceph.io/v1alpha1
kind: Driver
metadata:
  name: rbd.csi.ceph.com
  namespace: <operator-namespace> 
spec:
  controllerPlugin:
    hostNetwork: false
```

let use `useHostNetwork` or `hostNetwork` like we have in pod spec
